graveler/branch_locker.go

```diff
@@ -1,96 +1,82 @@
 package graveler
 
 import (
+	"context"
 	"fmt"
-	"sync"
+	"hash/fnv"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/treeverse/lakefs/db"
 )
 
-// branchLocker enforces the branch locking logic
+// BranchLocker enforces the branch locking logic
 // The logic is as follows:
 // allow concurrent writers to branch
 // metadata update commands will wait until all current running writers are done
 // while metadata update is running or waiting to run, writes and metadata update commands will be blocked
-type branchLocker struct {
-	locker   sync.Locker
-	c        *sync.Cond
-	branches map[string]*branchLockerData
-}
-
-type branchLockerData struct {
-	writers        int
-	metadataUpdate bool
-}
-
-func NewBranchLocker() branchLocker {
-	m := sync.Mutex{}
-	return branchLocker{
-		locker:   &m,
-		c:        sync.NewCond(&m),
-		branches: make(map[string]*branchLockerData),
-	}
+type BranchLocker struct {
+	db db.Database
 }
 
-// AquireWrite returns a cancel function to release if write is currently available
-// returns ErrBranchLocked if metadata update is currently in progress
-func (l *branchLocker) AquireWrite(repositoryID RepositoryID, branchID BranchID) (func(), error) {
-	key := formatBranchLockerKey(repositoryID, branchID)
-	l.locker.Lock()
-	defer l.locker.Unlock()
-	data := l.branches[key]
-	if data == nil {
-		data = &branchLockerData{}
-		l.branches[key] = data
-	} else if data.metadataUpdate {
-		return nil, ErrBranchLocked
+func NewBranchLocker(db db.Database) *BranchLocker {
+	return &BranchLocker{
+		db: db,
 	}
-	data.writers++
-	return func() { l.releaseWrite(key) }, nil
 }
 
-func (l *branchLocker) releaseWrite(key string) {
-	l.locker.Lock()
-	defer l.locker.Unlock()
-	branchData := l.branches[key]
-	branchData.writers--
-	if branchData.writers > 0 {
-		return
-	}
-	if branchData.metadataUpdate {
-		l.c.Broadcast()
-	} else {
-		delete(l.branches, key)
-	}
-}
-
-// AquireMetadataUpdate returns a cancel function to release if metadata update is currently available
-// Will wait until all current writers are done
-// returns ErrBranchLocked if metadata update is currently in progress
-func (l *branchLocker) AquireMetadataUpdate(repositoryID RepositoryID, branchID BranchID) (func(), error) {
-	key := formatBranchLockerKey(repositoryID, branchID)
-	l.locker.Lock()
-	defer l.locker.Unlock()
-	data := l.branches[key]
-	if data == nil {
-		data = &branchLockerData{}
-		l.branches[key] = data
-	} else if data.metadataUpdate {
-		// allow just one metadata update at a time
-		return nil, ErrBranchLocked
-	}
-	// wait until all writers will leave
-	data.metadataUpdate = true
-	for data.writers > 0 {
-		l.c.Wait()
-	}
-	return func() { l.releaseMetadataUpdate(key) }, nil
+// Writer try to lock as writer, using postgres advisory lock for the span of calling `lockedCB`.
+// returns ErrLockNotAcquired in case we fail to acquire the lock or commit is in progress
```

Comment:
```suggestion
// Writer tries to acquire the lock as a writer. On success, it calls `lockedCB` and then releases the lock.
// Returns ErrLockNotAcquired in case we fail to acquire the lock.
```

Reply:
mixed your comment with ariels's comment